### ISSUE TITLE
Keyref link breaks if target topic contains keyref

### ISSUE THREAD

**Issue by eerohele** [dita-ot/dita-ot#2134]
`develop` branch. [Sample files](https://github.com/eerohele/dita-ot-issues/tree/master/fixtures/2134).

Easier to show than tell: the first link in `topic1.dita` is broken, the second one works. If you remove `<keyword keyref="keyword"/>` from `topic2.dita`, the first link starts working.

It looks like the `keyref` stage creates a file called `topic2-1.dita` in `temp`. It is an exact clone of `topic2.dita`. The subsequent preprocessing steps then use `topic2-1.dita` instead of `topic2.dita`. Didn't get further than that yet.

**EDIT**: This is a regression, by the way — doesn't occur on 2.0.

**Comment by eerohele** [comment]
Seems to be somehow related to [`KeyrefModule.java#L169-L171`](https://github.com/dita-ot/dita-ot/blob/develop/src/main/java/org/dita/dost/module/KeyrefModule.java#L169-L171) and [`KeyrefModule.java#L226-L231`](https://github.com/dita-ot/dita-ot/blob/develop/src/main/java/org/dita/dost/module/KeyrefModule.java#L226-L231), which track topics that use keyrefs, but I'm not up-to-date with keyscopes, so I don't know what exactly is going on.

**Comment by eerohele** [comment]
Quite the mind twister, this one...

Still trying to get the full picture, but:

1. `KeyrefModule.java` walks through all topicrefs in the input map.
2. It checks whether topicrefs are used more than once.
3. For every topic that's used more than once, it creates a clone.

Subsequent processing then uses the clone instead of the original topic, which causes the broken links.

The input map to the keyref module looks like this (some noise omitted):

```xml

<map id="root">
  <title class="- topic/title ">DITA Map</title>
  <keydef class="+ map/topicref mapgroup-d/keydef ditaot-d/keydef " format="ditamap" href="keydef.ditamap" keys="variables" processing-role="resource-only"/>
  <submap class="+ map/topicref mapgroup-d/topicgroup ditaot-d/submap " dita-ot:orig-class="+ map/topicref mapgroup-d/keydef " dita-ot:orig-format="ditamap" dita-ot:orig-href="keydef.ditamap" processing-role="resource-only">
    <keydef class="+ map/topicref mapgroup-d/keydef " href="topic2.dita" keys="topic2" processing-role="resource-only"/>
    <keydef class="+ map/topicref mapgroup-d/keydef " href="topic3.dita" keys="topic3" processing-role="resource-only"/>
    <keydef class="+ map/topicref mapgroup-d/keydef " keys="keyword" processing-role="resource-only">
      <topicmeta class="- map/topicmeta ">
        <keywords class="- topic/keywords ">
          <keyword class="- topic/keyword ">foobar</keyword>
        </keywords>
      </topicmeta>
    </keydef>
  </submap>
  <topicref class="- map/topicref " href="topic1.dita"/>
  <topicref class="- map/topicref " href="topic2.dita"/>
  <topicref class="- map/topicref " href="topic3.dita"/>
</map>
```

`topic2.dita` is referenced twice: in `<keydef keys="topic2" href="topic2.dita">` and in `<topicref href="topic2.dita">`. This causes the `topic2-1.dita` clone to be created.

Are the clones created because of this passage in the [spec](http://docs.oasis-open.org/dita/dita/v1.3/csprd01/part3-all-inclusive/archSpec/base/processing-key-references-general.html#processing_key_references)?

>If a topic that contains key references is reused in multiple key scopes within a given root map such that its references resolve differently in each use context, processors MUST produce multiple copies of the source topic in resolved output for each distinct set of effective key definitions that are referenced by the topic. In such cases, authors can use the @copy-to attribute to specify different source URIs for each reference to a topic.

If so, I guess the problem is that current code creates clones even within a single key scope?

I guess one option would also be to ignore resource-only topicrefs when keeping track of topic uses. So something like:

```diff
diff --git a/src/main/java/org/dita/dost/module/KeyrefModule.java b/src/main/java/org/dita/dost/module/KeyrefModule.java
index 978bbcf..dbbf0a9 100644
--- a/src/main/java/org/dita/dost/module/KeyrefModule.java
+++ b/src/main/java/org/dita/dost/module/KeyrefModule.java
@@ -159,8 +159,11 @@ final class KeyrefModule extends AbstractPipelineModuleImpl {
         if (hrefNode == null) {
             hrefNode = elem.getAttributeNode(ATTRIBUTE_NAME_HREF);
         }
+
+        Attr processingRole = elem.getAttributeNode(ATTRIBUTE_NAME_PROCESSING_ROLE);
+
         for (final KeyScope s: ss) {
-            if (hrefNode != null) {
+            if (hrefNode != null && (processingRole == null || ATTR_PROCESSING_ROLE_VALUE_NORMAL.equals(processingRole.getValue()))) {
                 final URI href = stripFragment(job.getInputMap().resolve(hrefNode.getValue()));
                 final FileInfo fi = job.getFileInfo(href);
                 if (fi != null && fi.hasKeyref) {
```

After all, the I don't think it's necessary to create clones for resource-only topics.

**Comment by drmacro** [comment]
I think the most accurate algorithm depends on whether or not resource-only keys are referenced from any normal-role topics. That is, processors may choose to treat resource-only topicrefs as normal-role in that case. 

In the case of HTML output one can argue that it's appropriate to generate HTML pages for otherwise resource-only key definitions if those keys are referenced from other normal-role topics, either via direct xref or through a relationship table. In that case, resource-only topicrefs count toward the use count, but only if the key is referenced directly and not indirectly from other normal-role topicrefs. I personally think that this behavior is not the correct behavior--the processor should report references to resource-only keys and treat the key as unresolvable for the purposes of xrefs and reltables. But I also realize that that's a rather draconian response and is different from what the HTML transform has done historically. 

[Note that navigation references to resource-only keys are only slightly better than direct URI references to the topics because the resource-only topicrefs do not establish any use context for the topics, so you have the same lack-of-context problem there as with direct URI references. Only references to normal-role topicrefs establish an unambiguous use context. Thus all navigation keyrefs should be to normal-role topicrefs, not resource-only topicrefs.]

In the case of PDF output it doesn't really make sense to treat resource-only topics as normal since there's no well-defined place to include them in the output. One could define a policy but I can't imagine a case where you would want that behavior for production use (it could make sense in some sort of review draft or something where you'd rather report the error but include the topic in the reviewable artifact).

So clearly different processors need to behave differently and probably provide runtime options as to how they handle this case.

I think this implies an algorithm like:

1. Resolve the directly-referenced map tree
2. Apply branch filtering and key scope construction to construct the effective map and key spaces
3. Evaluate all normal-role topics and reltables to determine if there are any direct navigation references to resource-only keys. 
4. If the policy is "treat resource-only as normal for navigation" then update the relevant key definition records from resource-only to normal
5. Determine use counts within scopes
6. Generate copies as required

The preprocessing must already be looking at all topics to determine what key references they use so step 3 is not really new processing, but it might come at a different point in the current preprocessing chain.

**Comment by eerohele** [comment]
So I might be completely off-track here, but is the solution to keep track of the key scopes that each topicref (identified by its absolute URI) is used in and add the `-1` suffix to the filename only if the topicref is used in more than one keyscope? So something along the lines of:

```diff
diff --git a/src/main/java/org/dita/dost/module/KeyrefModule.java b/src/main/java/org/dita/dost/module/KeyrefModule.java
index c8e2dea..a65138a 100644
--- a/src/main/java/org/dita/dost/module/KeyrefModule.java
+++ b/src/main/java/org/dita/dost/module/KeyrefModule.java
@@ -48,6 +48,7 @@ final class KeyrefModule extends AbstractPipelineModuleImpl {
     private String transtype;
     final Set<URI> normalProcessingRole = new HashSet<>();
     final Map<URI, Integer> usage = new HashMap<>();
+    final Map<URI, Set<KeyScope>> keyScopesByURI = new HashMap<URI, Set<KeyScope>>();
     private TopicFragmentFilter topicFragmentFilter;
 
     /**
@@ -166,7 +167,7 @@ final class KeyrefModule extends AbstractPipelineModuleImpl {
                 if (fi != null && fi.hasKeyref) {
                     res.add(processTopic(fi, s));
                     final Integer used = usage.get(fi.uri);
-                    if (used > 1) {
+                    if (used > 1 && keyScopesByURI.get(fi.uri).size() > 1) {
                         hrefNode.setValue(addSuffix(toURI(hrefNode.getValue()), "-" + (used - 1)).toString());
                     }
                 }
@@ -186,7 +187,15 @@ final class KeyrefModule extends AbstractPipelineModuleImpl {
         final Integer used = usage.containsKey(f.uri) ? usage.get(f.uri) + 1 : 1;
         usage.put(f.uri, used);
 
-        if (used > 1) {
+        if (keyScopesByURI.get(f.uri) == null) {
+            Set<KeyScope> scopes = new HashSet<KeyScope>();
+            scopes.add(scope);
+            keyScopesByURI.put(f.uri, scopes);
+        } else {
+            keyScopesByURI.get(f.uri).add(scope);
+        }
+
+        if (used > 1 && keyScopesByURI.get(f.uri).size() > 1) {
             final URI out = addSuffix(f.uri, "-" + (used - 1));
             final FileInfo fo = new FileInfo.Builder(f).uri(out).build();
             // TODO: Should this be added when content is actually generated?
```

**Comment by eerohele** [comment]
Fixed via #2304.
